appinventor/components/src/com/google/appinventor/components/runtime/HorizontalScrollArrangement.java

```diff
@@ -25,10 +35,129 @@
     "scrollable.",
     category = ComponentCategory.LAYOUT)
 @SimpleObject
-public class HorizontalScrollArrangement extends HVArrangement {
+public class HorizontalScrollArrangement extends HVArrangement implements OnScrollChangedListener {
+
+  private int oldScrollX = 0;
+
   public HorizontalScrollArrangement(ComponentContainer container) {
     super(container, ComponentConstants.LAYOUT_ORIENTATION_HORIZONTAL,
       ComponentConstants.SCROLLABLE_ARRANGEMENT);
+    getView().getViewTreeObserver().addOnScrollChangedListener(this);
+  }
+
+  private int px2dx(int px) {
+    return Math.round(px * container.$form().deviceDensity());
+  }
+
+  private int dx2px(int dx) {
+    return Math.round(dx / container.$form().deviceDensity());
+  }
+
+  private HorizontalScrollView getScrollView() {
+    return (HorizontalScrollView) getView();
+  }
+
+  @Override
+  public void onScrollChanged() {
+    int scrollX = dx2px(getView().getScrollX());
+    if (scrollX < 0) {
+      scrollX = 0;
+    }
+    if (oldScrollX != scrollX) {
+      ScrollChanged(scrollX);
+      oldScrollX = scrollX;
```

Comment:
I would update `oldScrollX` before dispatching the event. Otherwise, if there is a thrown exception in the handler the `oldScrollX` field might not be updated. Alternatively, wrap in a `try-finally` block.